Jennyf/adfx fix

jennyf19 opened the pull request:
#1693

jennyf19 committed:
fix adfs literal being dropped and improve unit tests

jennyf19 committed:
more fixes

henrik-me commented on src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/Authenticator.cs:
```diff
@@ -109,18 +109,24 @@ public async Task UpdateFromTemplateAsync(RequestContext requestContext)
 
-                // The authority could be https://{AzureAD host name}/{tenantid} OR https://{Dsts host name}/dstsv2/{tenantid}
+                // The authority could be https://{AzureAD host name}/{tenantid} OR https://{Dsts host name}/dstsv2/{tenantid} OR
+                // https://{adfs host name}/adfs/{tenantid}
```
>https://{adfs host name}/adfs/{tenantid} [](start = 19, length = 40)

I thought ADFS wouldn't necessarily have a tenantid. we probably have to support both? #Resolved

keystroke replied:
ADFS will never have a tenantId. ARM etc. will advertise and use /adfs for the OIDC endpoint which is correct and works. KeyVault has code patterns from AAD where they take a "base" authority URI in their code and add a tenantId as a suffix to that, so they end-up using /adfs/tenantId in the endpoints. They can't fix the code to directly support the ADFS authority URI so the ask is for ADAL to support passing-in the invalid authority URI and correcting it under-the-hood. Note that this same ask will be levied against MSAL as well. It seems there is a lot of this code from Azure Services that isn't going to be precise about the OIDC protocol and how authority URIs work, so it may be reasonable to have ADAL and MSAL "tolerate" such code and internally look for "/adfs" on the first segment and make the necessary transformations or w/e for ADFS. In #1653 we discussed this a bit more and how the MSAL design for .WithAdfsAuthority is also problematic for these same kinds of code patterns. Web applications that integrate with identity need to be built supporting both identity providers, but are we going to make end-users parse the authority URI again themselves in their code and branch to call .WithAuthority or .WithAdfsAuthority? This means ADAL / MSAL internally understanding authority URIs that begin with /adfs and "knowing" that it is an ADFS-backed IDP and whatever that needs to mean to make things work. #Resolved

jennyf19 replied:
sounds good. thanks for clarifying. I will remove the {tenantid} here, but it doesn't impact any of the changes.

---
In reply to: [378542753](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1694#discussion_r378542753) [](ancestors = 378542753)

henrik-me replied:
nice, as discussed. we should have tests for both scenarios.

jennyf19 replied:
yes...will add.

henrik-me commented on tests/Test.ADAL.NET.Unit.net45/InstanceDiscoveryTests.cs:
```diff
@@ -173,7 +173,7 @@ public async Task TestInstanceDiscovery_WhenAuthorityIsAdfs_ShouldNotDoInstanceD
             {
                 var serviceBundle = ServiceBundle.CreateWithCustomHttpManager(httpManager);
-                var authenticator = new Authenticator(serviceBundle, "https://login.contoso.com/adfs", false);
+                var authenticator = new Authenticator(serviceBundle, "https://login.contoso.com/adfs/" + AdalTestConstants.SomeTenantId, false);
```
>/ [](start = 100, length = 1)

does this mean our code doesn't work if there is no / at the end?

henrik-me commented on tests/Test.ADAL.NET.Unit.net45/InstanceDiscoveryTests.cs:
```diff
@@ -173,7 +173,7 @@ public async Task TestInstanceDiscovery_WhenAuthorityIsAdfs_ShouldNotDoInstanceD
             using (var httpManager = new MockHttpManager())
             {
                 var serviceBundle = ServiceBundle.CreateWithCustomHttpManager(httpManager);
-                var authenticator = new Authenticator(serviceBundle, "https://login.contoso.com/adfs", false);
```
>https://login.contoso.com/adfs [](start = 70, length = 30)

i believe this should still be possible #Resolved

jennyf19 replied:
yes it is. i removed the tenantId. 

---
In reply to: [378037207](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1694#discussion_r378037207) [](ancestors = 378037207)

henrik-me reviewed (changes requested):
:clock1:

jennyf19 committed:
remove tenantId from tests

jennyf19 commented:
ok...addressed the comments.

---
In reply to: [357184636](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1694#issuecomment-357184636) [](ancestors = 357184636)

henrik-me reviewed (approved):
:shipit:

jennyf19 commented:
PR build passed here: https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=473498&view=results

jkapil commented:
The dSTS V2 and ADFS have fixed tenant Id. In the context of ADFS tenantId is not adfs , it is a GUID. However when ADAL needs to create URL for getting tokens, it needs to ignore tenantId for ADFS (until ADFS fix comes - which could be never) and hard-code the outbound URL to {host}/adfs. 

Current commit to revert the regression scoped to ADFS scenario will also work.
